consensus/safety-rules/src/safety_rules.rs

```diff
@@ -148,16 +147,19 @@ impl SafetyRules {
     /// @TODO if public key does not match private key in validator set, access persistent storage
     /// to identify new key
     pub fn update(&mut self, qc: &QuorumCert) {
+        if qc.parent_block().round() > self.persistent_storage.preferred_round() {
+            self.persistent_storage
+                .set_preferred_round(qc.parent_block().round());
+        }
+    }
+
+    /// Notify the safety rules about the new epoch start.
+    pub fn start_new_epoch(&mut self, qc: &QuorumCert) {
```

Comment:
this probably needs to be more specific, we either ensure qc.certified_block().round() == 0 or we call update(qc) inside
also I thought we need to pass the end-epoch QC to safety rule and verify it(of course no in this PR)

Reply:
Oh right makes sense.

Reply:
Maybe move around TODOs to follow this function?
In all actuality safety rules needs to see the epoch ending commit that contains the committed validator set, so I think that is still likely to happen in update unless we want a more explicit API for handling that QC / ledger info.